ethereum/core/src/main/java/org/hyperledger/besu/ethereum/worldstate/WorldStateStorageCoordinator.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright Hyperledger Besu Contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.worldstate;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.ethereum.trie.bonsai.storage.BonsaiWorldStateKeyValueStorage;
+import org.hyperledger.besu.ethereum.trie.forest.storage.ForestWorldStateKeyValueStorage;
+
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+
+public class WorldStateStorageCoordinator {
+  private final WorldStateKeyValueStorage worldStateKeyValueStorage;
+
+  public WorldStateStorageCoordinator(final WorldStateKeyValueStorage worldStateKeyValueStorage) {
```

Comment:
👍  to the coordinator concept.  I like that this is the central place to add/remove storage formats and their interfaces.

the only bits I do not care for are where we double-down on forest and bonsai storage modes in the signatures, the apply and consume ForStrategy methods basically.  I think we could do some lambda magic to limit the bonsai/forest stuff in these methods, like: 

```
public <RESPONSE, T extends WorldStateKeyValueStorage> RESPONSE applyForStrategy(
      final Function<T, RESPONSE> strategy) {
 ```

Reply:
Why I am adding Bonsai and Forest in the signature is because I no longer want a common interface between Bonsai and Forest. The coordinator will play the role of redirection and do the cast to simplify other calls. So, as I don't have a common interface, I can no longer create a single strategy interface that I pass to the coordinator. I must therefore pass both and the coordinator calls the right one. I can do that but I will have to add the cast logic in other calls and in order to simplify this call I will add complexity to the other calls 

If I recreate an interface, it's doing what we had before and I specifically wanted to avoid that. Knowing that Forest will be eliminated soon, this should also remove the coordinator and this mechanism. But maybe I didn't understand your idea.

Reply:
This is what I was suggesting:

https://github.com/garyschulte/besu/commit/aab618747dfa61abbe9360de6703991e98366cb1

I didn't remove all the applyWhen, consumeFor, etc but I think we can simplify the coordinator a lot this way

Reply:
could you confirm on this PR that it's ok for you to merge with the current implementation ?

Reply:
Confirmed 👍